### DISCUSSION TITLE
How to get the uv positions of vertices in a torch tensor effectively?

### DISCUSSION THREAD

**Discussion by aGIToz** [facebookresearch/pytorch3d#588] [Q&A]
Moved from [issue-586](https://github.com/facebookresearch/pytorch3d/issues/586)

**Comment by bottler** [comment]
The texture data (faces.texture_idx and aux.verts_uvs) makes a link between each triangular face of the mesh and a triangle of UV space. But there is no guarantee that if multiple faces intersect at the same vertex then the corresponding corners of those faces at that point in space will have the same UV values. So getting the "uv positions of vertices" isn't well defined.

 If you aren't concerned how these ambiguities are resolved, and you don't care about vertices which aren't used at all, then you can get a uv coordinate for each vertex with something like the following. Note that the desired behavior closely matches the idea of an indexed assignment in torch/numpy; after realizing this it is straightforward to come up with the code.

```
V, _ = verts
verts_to_uv_index = torch.zeros(V, dtype=torch.int64)
verts_to_uv_index[faces.verts_index.flatten()]=faces.textures_index.flatten()  # <- This is where the carelessness is
verts_to_uvs = aux.verts_uvs[verts_to_uv_index]
```
